[#22] 회원 탈퇴 기능 구현

thinker-94 opened the pull request:
- pom.xml aop 의존성 추가
- UserController UserInfoController, LoginController로 분리
- 유저의 세션을 확인하는 aspect 추가
- 인증 AOP 적용을 명시 하기 위한 커스텀@Authorized 추가
- 유저 삭제 쿼리문 추가

Commit:
[#22] 회원 탈퇴 기능 구현

- pom.xml aop 의존성 추가
- UserController UserInfoController, LoginController로 분리
- 유저의 세션을 확인하는 aspect 추가
- 인증 AOP 적용을 명시 하기 위한 커스텀@Authorized 추가
- 유저 삭제 쿼리문 추가

Commit:
[#30] 회원 수정 기능 구현, 유저 탈퇴 기능 수정

- 유저 수정 기능 추가
- SessionCheckAspect.java의 session 삭제하는 코드 잘못 넣어서 삭제, 유저 탈퇴 기능에 추가
- 메소드명 일관성 있도록 리팩토링

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/aspect/SessionCheckAspect.java:
```diff
@@ -0,0 +1,47 @@
+    String userId = (String) joinPoint.getArgs()[0];
+    String userSession = (String) session.getAttribute(USER_SESSION_KEY);
+
+    if (userSession.equals(userId)){
```
`userSession`이 null이면 NPE가 나지 않을까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/aspect/SessionCheckAspect.java:
```diff
@@ -0,0 +1,47 @@
+
+    if (userSession.equals(userId)){
+        session.invalidate();
+        System.out.println("삭제 완료");
```
`System.out.println`은 쓰지 않는게 좋습니다. 왜일까요?

thinker-94 replied:
콘솔로 출력하기 때문에 성능에 문제가 생길 수 있습니다.

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/aspect/SessionCheckAspect.java:
```diff
@@ -0,0 +1,47 @@
+    String userSession = (String) session.getAttribute(USER_SESSION_KEY);
+
+    if (userSession.equals(userId)){
+        session.invalidate();
```
왜 로그아웃을 해줄까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/aspect/SessionCheckAspect.java:
```diff
@@ -0,0 +1,47 @@
+
+@Aspect
+@Component
+public class SessionCheckAspect {
```
사용자 관점에서는 `SessionCheck`하고 `LoginCheck` 중에 어떤게 더 역할을 명확하게 파악하기 좋을까요?

thinker-94 replied:
후자가 더 좋겠네요

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/LoginController.java:
```diff
@@ -0,0 +1,42 @@
+  private UserService userService;
+
+  @PostMapping
+  public ResponseEntity<UserDto> login(UserDto userDto,
```
이 PR의 제목은 회원탈퇴인데 왜 로그인 작업이 껴있을까요?

thinker-94 replied:
조심하겠습니다 !

f-lab-dev commented on src/main/resources/query/user/userQuery.xml:
```diff
@@ -4,12 +4,17 @@
-        SELECT user_password FROM user
-        WHERE user_id = (#{userId})
+        SELECT userPassword FROM user
+        WHERE userId = (#{userId})
```
이거도 저번 PR에서 지나친것같은데 괄호는 왜 사용하셨을까요?

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/controller/UserInfoController.java:
```diff
@@ -35,25 +30,16 @@
+  @DeleteMapping(value = "{userId}")
+  public ResponseEntity<String> delete(@PathVariable String userId) {
+    userService.deleteUser(userId);
+    return ResponseEntity.status(HttpStatus.OK).body(userId);
```
꼭 클라이언트한테 userId를 보내줘야할까요?

Commit:
[#22] 코드 리뷰 반영

- 유저 세션 NULL 예외처리
- 예외 커스텀 클래스 작성
- LoginCheckAspect로 SessionCheckAspect 클래스 이름 변경

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/aspect/LoginCheckAspect.java:
```diff
@@ -36,9 +37,11 @@ public void checkSessionValid(JoinPoint joinPoint) {
     String userSession = (String) session.getAttribute(USER_SESSION_KEY);
 
+    if (userSession == null) {
+      throw new SessionNullPointException();
```
이 예외를 받는 사용자는 세션이라는 단어가 와닿을까요? 로그인이 안되어있다는 메세지가 더 와닿을까요?

Commit:
[#22] 코드리뷰 반영

SessionNullPointException -> LoginRequiredException

Commit:
Merge confilt resolved

f-lab-dev commented on src/main/java/com/youngsuk/bookstore/exception/LoginRequiredException.java:
```diff
@@ -3,8 +3,8 @@
-    public SessionNullPointException() {
+public class LoginRequiredException extends ResponseStatusException {
+    public LoginRequiredException() {
         super(HttpStatus.BAD_REQUEST, "User Session Null Exception");
```
메세지에 유저 세션이 null이다라는 메세지는 그대로로 보입니다~

f-lab-dev replied:
이거는 왜 resolve 처리 하셨나요??

thinker-94 committed:
Merge pull request #31 from f-lab-edu/feature/10-20-23-27-22-30

[#30] 회원 수정 기능 구현, 유저 탈퇴 기능 수정
